### PR TITLE
DEV2-2899: add the chat webview for alpha users 

### PR DESCRIPTION
**Pull request by yairco1990**

### CODE REVIEW

File: chat/package.json [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  "name": "chat",
+  "version": "0.1.0",
+  "private": true,
+  "dependencies": {
+    "axios": "^1.4.0",
+    "constate": "^3.3.2",
+    "lodash": "^4.17.21",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+    "react-scripts": "5.0.1",
```
Comment (amirbilu) [inline chat/package.json:11]: dev

File: chat/package.json [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    "react-syntax-highlighter": "^15.5.0",
+    "styled-components": "^5.2.1",
+    "ts-loader": "^9.4.1",
```
Comment (amirbilu) [inline chat/package.json:14]: dev

File: chat/src/components/general/ChatInput.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+          onSubmit(message);
+          setMessage("");
+        }}
+      >
+        <RightArrowPadding>
+          <RightArrowIcon />
+        </RightArrowPadding>
+      </RightArrowContainer>
+      <Textarea
+        ref={textareaRef}
```
Comment (amirbilu) [inline chat/src/components/general/ChatInput.tsx:34]:
```suggestion
        ref={textareaRef=>textareaRef && textareaRef.focus()}
```

File: chat/src/components/general/ChatInput.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+const Wrapper = styled.div`
+  position: relative;
+  border-top: 1px solid #433f47;
+`;
+
+const RightArrowContainer = styled.div`
+  position: absolute;
+  right: 8px;
+  top: 9px;
```
Comment (amirbilu) [inline chat/src/components/general/ChatInput.tsx:66]: Padding maybe? not sure

File: chat/src/components/general/ChatInput.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  padding: 4px;
+`;
+
+const Textarea = styled.textarea`
+  width: 100%;
+  height: 100%;
+  background-color: var(--vscode-editor-background);
+  color: var(--vscode-editor-foreground);
+  outline: none !important;
+  border: none !important;
```
Comment (amirbilu) [inline chat/src/components/general/ChatInput.tsx:90]:
```suggestion
&&& {
  outline: none;
  border: none;
  }
```

File: chat/src/components/message/BotErrorMessage.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+export function BotErrorMessage({ onRegenerate }: Props): React.ReactElement {
+  const { id: conversationId } = useConversationContext();
+  return (
+    <MessageContextProvider
+      isError
+      message={{
+        conversationId,
+        isBot: true,
+        text:
```
Comment (amirbilu) [inline chat/src/components/message/BotErrorMessage.tsx:21]: Link to support page?

File: chat/src/components/message/BotIsTypingMessage.tsx [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+      return;
+    }
+    if (!isLoading) {
+      onFinish({
+        id: messageId,
+        content: data,
+      });
+      return;
+    }
+  }, [messageId, data, isLoading, error, onTextChange, onError, onFinish]);
```
Comment (amirbilu) [inline chat/src/components/message/BotIsTypingMessage.tsx:51]: Feels like every part here deserves it's own `useEffect` (with dedicated dep array)

File: chat/src/components/message/BotMessageHeader.tsx [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  display: flex;
+  justify-content: space-between;
+`;
+
+const RateIconsContainer = styled.div`
+  display: flex;
+  align-items: center;
+  & > *:not(:last-child) {
+    margin: 0 0.5rem;
+  }
```
Comment (amirbilu) [inline chat/src/components/message/BotMessageHeader.tsx:82]: `gap`?

File: chat/src/components/message/BotMessageHeader.tsx [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+  ${({ selectedThumbs }) =>
+    selectedThumbs
+      ? ""
+      : css`
+          &:hover {
+            path {
+              fill: #e0e0e0;
+            }
+          }
+        `};
```
Comment (amirbilu) [inline chat/src/components/message/BotMessageHeader.tsx:108]:
```suggestion
    !selectedThumbs
      && "
          &:hover {
            path {
              fill: #e0e0e0;
            }
          }
        "};
```

File: chat/src/components/message/MessageContainer.tsx [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+import styled from "styled-components";
+import { useMessageContext } from "../../hooks/useMessageContext";
+
+type Props = {
+  children: React.ReactNode;
+};
+
+export function MessageContainer({
+  children,
+  ...props
```
Comment (amirbilu) [inline chat/src/components/message/MessageContainer.tsx:11]: ?

File: chat/src/components/message/MessageContainer.tsx [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import React from "react";
+import styled from "styled-components";
+import { useMessageContext } from "../../hooks/useMessageContext";
+
+type Props = {
+  children: React.ReactNode;
+};
```
Comment (amirbilu) [inline chat/src/components/message/MessageContainer.tsx:7]: import { PropsWithChildren } from "react"

File: chat/src/components/message/MessageContent.tsx [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+type MessageContentTypeProps = {
+  textSegments: MessageSegment[];
+};
+
+export function MessageContentType({
+  textSegments,
+}: MessageContentTypeProps): React.ReactElement {
+  const { conversationMessages } = useChatState();
+  const { message } = useMessageContext();
+  const spanRef = useRef<HTMLDivElement | null>(null);
```
Comment (amirbilu) [inline chat/src/components/message/MessageContent.tsx:43]: ?

File: chat/src/components/general/ChatInput.tsx (hunk 4 above)
Comment (yairco1990) [reply chat/src/components/general/ChatInput.tsx:66]: I think it's ok. it should be painted on the textarea

File: chat/src/components/views/ConversationView.tsx [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        </>
+      </ChatMessagesHolder>
+    </Wrapper>
+  );
+}
+
+const Wrapper = styled.div`
+  padding-bottom: 0.4rem;
+  overflow-y: auto;
+  height: 100%;
```
Comment (amirbilu) [inline chat/src/components/views/ConversationView.tsx:121]:
```suggestion
```

File: chat/src/components/message/BotErrorMessage.tsx (hunk 6 above)
Comment (yairco1990) [reply chat/src/components/message/BotErrorMessage.tsx:21]: this is just a message for the analytics, but I'll add it.

File: chat/src/components/views/ConversationView.tsx (hunk 13 above)
Comment (amirbilu) [reply chat/src/components/views/ConversationView.tsx:121]: (if parent flex direction is column)

File: chat/src/components/views/HistoryView.tsx [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+          )}
+        </Top>
+        {!hasConversations && <HistoryEmptyState />}
+        {Object.values(conversations)
+          .filter((conversation) => conversation.messages.length > 0)
+          .sort(
+            (c1, c2) =>
+              Number(c2.messages[c2.messages.length - 1].timestamp) -
+              Number(c1.messages[c1.messages.length - 1].timestamp)
+          )
```
Comment (amirbilu) [inline chat/src/components/views/HistoryView.tsx:34]: `useMemo`

File: chat/src/hooks/ExtensionCommunicationProvider.tsx [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    function handleResponse(event: MessageEvent) {
+      const message: ExtensionMessage<any> = event.data;
+
+      // if this is a response to a request, resolve the corresponding promise
+      if (pendingRequests.has(message.id)) {
+        const pendingRequest = pendingRequests.get(message.id);
+        pendingRequest?.resolve(message.payload);
+        pendingRequests.delete(message.id);
+      }
+    }
```
Comment (amirbilu) [inline chat/src/hooks/ExtensionCommunicationProvider.tsx:56]: can be moved out of the component (unless if missing something)

File: chat/src/hooks/useChatBotQueryData.ts [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+  conversationId: string;
+  messageId: string;
+  editorContext?: EditorContext;
+};
+
+export function useChatBotQueryData(): ChatBotQueryData | null {
+  const userDetails = useUserDetails();
+  const { editorContext, isEditorContextChanged } = useEditorContext();
+  const { conversationMessages, currentConversation } = useChatState();
+  const messageId = useMemo(() => uuidv4(), []);
```
Comment (amirbilu) [inline chat/src/hooks/useChatBotQueryData.ts:19]: `useMemo` used for optimization not for consistency. It's not promised to not run twice. In other words - move the `messageId` out of the component

File: chat/src/hooks/useChatState.tsx [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+    setConversationMessages([]);
+  }, [setConversationMessages]);
+
+  const setCurrentConversationData = useCallback(
+    (conversation: ChatConversation) => {
+      setCurrentConversation(conversation);
+      setConversationMessages(conversation.messages);
+    },
+    [setCurrentConversation, setConversationMessages]
+  );
```
Comment (amirbilu) [inline chat/src/hooks/useChatState.tsx]: `useReducer` feel more appropriate here, but I understand it will be hard to refactor now

Comment (yairco1990) [reply chat/src/hooks/useChatState.tsx]: I'll find some time to change this

File: chat/src/utils/events.ts [hunk 18]
```diff
@@ -0,0 +1,187 @@
+import { snakeCase } from "lodash";
```
Comment (amirbilu) [inline chat/src/utils/events.ts:1]: 1980

File: chat/package.json [hunk 19]
```diff
@@ -0,0 +1,61 @@
+{
+  "name": "chat",
+  "version": "0.1.0",
+  "private": true,
+  "dependencies": {
+    "axios": "^1.4.0",
```
Comment (amirbilu) [inline chat/package.json:6]: I see you're not using this

Comment (yairco1990) [reply chat/package.json:6]: I tried to use it for the stream, but it was bad. removed :)

File: package.json [hunk 20]
```diff
@@ -100,14 +100,16 @@
     "compile": "webpack --mode development --watch",
     "build": "webpack --mode development",
     "build:enterprise": "webpack --mode development --env enterprise",
+    "chat:build": "cd chat && yarn && yarn build",
     "test:copyassets": "ncp ./src/test/fixture ./out/test/fixture/",
-    "prettier": "prettier --write src/",
-    "prettier:check": "prettier --check src/",
+    "prettier": "prettier --write src/ chat/",
+    "prettier:check": "prettier --check src/ chat/",
```
Comment (amirbilu) [inline package.json:106]: mistake?

File: chat/src/utils/events.ts (hunk 18 above)
Comment (yairco1990) [reply chat/src/utils/events.ts:1]: changed it
https://www.npmjs.com/package/snakecase-keys

File: package.json [hunk 21]
```diff
@@ -100,14 +100,16 @@
...
     "lint": "eslint . --max-warnings 0",
     "lint:fix": "eslint . --fix",
     "pretest": "yarn clear-out && tsc && yarn test:copyassets",
     "test": "node ./out/test/runTest.js",
-    "vsce:package": "vsce package",
+    "vsce:prebuild": "yarn chat:build",
+    "vsce:package": "yarn vsce:prebuild && vsce package",
```
Comment (amirbilu) [inline package.json:112]: I'd expect `chat:build` will be part of `build`. Not of `vsce:package`

File: package.json [hunk 22]
```diff
@@ -197,20 +201,14 @@
...
         {
           "id": "tabnine-home",
           "name": "Quick Access",
           "when": "tabnine.tabnine-navigation-ready"
-        },
-        {
-          "type": "webview",
-          "id": "tabnine-notifications",
-          "name": "Notifications",
-          "when": "tabnine.notifications-ready"
```
Comment (amirbilu) [inline package.json:213]: Is this coordinated with product?

File: package.json (hunk 20 above)
Comment (yairco1990) [reply package.json:106]: why mistake?

File: src/tabnineChatWidget/ChatViewProvider.ts [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+      const uri = vscode.Uri.file(
+        path.join(this.extensionPath, "chat", "build", "static")
+      );
+      const webviewUri = webviewView.webview.asWebviewUri(uri);
+      return `${attribute}="${webviewUri}/`;
+    });
+    webviewView.webview.html = html;
+  }
+
+  setDevWebviewHtml(webviewView: WebviewView): void {
```
Comment (amirbilu) [inline src/tabnineChatWidget/ChatViewProvider.ts:122]: Nice

File: package.json (hunk 22 above)
Comment (yairco1990) [reply package.json:213]: with Nims. I'll check that again.

File: package.json (hunk 20 above)
Comment (amirbilu) [reply package.json:106]: this is the main `package.json`.. you're removing prettier from all other `src/`?

Comment (yairco1990) [reply package.json:106]: no, just added the /chat to it.
image

Comment (amirbilu) [reply package.json:106]: Oh sorry. Missed the space

**Review by amirbilu — APPROVED** [review]

File: src/tabnineChatWidget/tabnineChatWidgetWebview.ts [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+
+  context.subscriptions.push(
+    vscode.window.registerWebviewViewProvider(VIEW_ID, chatProvider, {
+      webviewOptions: {
+        retainContextWhenHidden: true, // keeps the state of the webview even when it's not visible
+      },
+    })
+  );
+
+  vscode.commands.registerCommand("tabnine.chat.submit-message", async () => {
```
Comment (dimacodota) [inline src/tabnineChatWidget/tabnineChatWidgetWebview.ts:35]: command is disposable, should be pushed to context.subscriptions

File: package.json [hunk 25]
```diff
@@ -100,14 +100,16 @@
...
-    "prettier": "prettier --write src/",
-    "prettier:check": "prettier --check src/",
+    "prettier": "prettier --write src/ chat/",
+    "prettier:check": "prettier --check src/ chat/",
     "lint": "eslint . --max-warnings 0",
     "lint:fix": "eslint . --fix",
     "pretest": "yarn clear-out && tsc && yarn test:copyassets",
     "test": "node ./out/test/runTest.js",
-    "vsce:package": "vsce package",
+    "vsce:prebuild": "yarn chat:build",
```
Comment (dimacodota) [inline package.json:111]: id run it as `prebuild`

File: chat/config-overrides.js [hunk 26]
```diff
@@ -0,0 +1,25 @@
+module.exports = function override(config, env) {
+  // Define our own filename
+  config.output.filename = "vscode.js";
```
Comment (dimacodota) [inline chat/config-overrides.js:3]: why this is the name?

File: .vscodeignore [hunk 27]
```diff
@@ -1,4 +1,7 @@
 **
+!chat/
+!chat/*
+!chat/build/**
```
[the remark this answers is not in the document]
Comment (yairco1990) [reply .vscodeignore:4]: this is the only way to do that (I checked a few different approaches). it includes only the build folder content along with the chat/ root folder files.
image

File: chat/config-overrides.js (hunk 26 above)
Comment (yairco1990) [reply chat/config-overrides.js:3]: this is just for dev purposes, we can change it

File: .vscodeignore (hunk 27 above)
[the remark this answers is not in the document]
Comment (yairco1990) [reply .vscodeignore:4]: we need pretty much all of this folder

**Review by dimacodota — APPROVED** [review]
lgtm

**Pull request merged**